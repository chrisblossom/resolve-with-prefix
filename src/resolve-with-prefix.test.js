/* @flow */

import path from 'path';
import { platform } from 'os';
import { normalizeRootPath } from '@chrisblossom/test-utils';

import ResolveWithPrefix from './resolve-with-prefix';

describe('ResolveWithPrefix', () => {
    const cwd = process.cwd();

    afterEach(() => {
        process.chdir(cwd);
    });

    it('resolves full preset without prefix set', () => {
        const dir = path.resolve(__dirname, '__sandbox__/app1/');
        process.chdir(dir);

        const resolve = new ResolveWithPrefix();

        const resolved = resolve('one-preset-test');

        const result = normalizeRootPath(resolved);
        expect(result).toMatchSnapshot();
    });

    it('resolves full preset with prefix set', () => {
        const dir = path.resolve(__dirname, '__sandbox__/app1/');
        process.chdir(dir);

        const resolve = new ResolveWithPrefix({ prefix: 'one-preset' });

        const resolved = resolve('one-preset-test');

        const result = normalizeRootPath(resolved);
        expect(result).toMatchSnapshot();
    });

    it('resolves leading preset with prefix', () => {
        const dir = path.resolve(__dirname, '__sandbox__/app1/');
        process.chdir(dir);

        const resolve = new ResolveWithPrefix({ prefix: 'one-preset' });

        const resolved = resolve('test');

        const result = normalizeRootPath(resolved);
        expect(result).toMatchSnapshot();
    });

    it('resolves other without preset', () => {
        const dir = path.resolve(__dirname, '__sandbox__/app1/');
        process.chdir(dir);

        const resolve = new ResolveWithPrefix();

        const resolved = resolve('other');

        const result = normalizeRootPath(resolved);
        expect(result).toMatchSnapshot();
    });

    it('resolves leading preset with prefix over matching module', () => {
        const dir = path.resolve(__dirname, '__sandbox__/app1/');
        process.chdir(dir);

        const resolve = new ResolveWithPrefix({ prefix: 'one-preset' });

        const resolved = resolve('other');

        const result = normalizeRootPath(resolved);
        expect(result).toMatchSnapshot();
    });

    it('strict:true, module: works', () => {
        const dir = path.resolve(__dirname, '__sandbox__/app1/');
        process.chdir(dir);

        const resolve = new ResolveWithPrefix({
            prefix: 'one-preset',
            strict: true,
        });

        const resolved = resolve('module:explicit-name-test');

        const result = normalizeRootPath(resolved);
        expect(result).toMatchSnapshot();
    });

    it('default strict:true, require module:', () => {
        const dir = path.resolve(__dirname, '__sandbox__/app1/');
        process.chdir(dir);

        const resolve = new ResolveWithPrefix({
            prefix: 'one-preset',
        });

        try {
            expect.assertions(1);

            resolve('explicit-name-test');
        } catch (error) {
            const result = normalizeRootPath(error);
            expect(result).toMatchSnapshot();
        }
    });

    it('strict:true, require module:', () => {
        const dir = path.resolve(__dirname, '__sandbox__/app1/');
        process.chdir(dir);

        const resolve = new ResolveWithPrefix({
            prefix: 'one-preset',
            strict: true,
        });

        try {
            expect.assertions(1);

            resolve('explicit-name-test');
        } catch (error) {
            const result = normalizeRootPath(error);
            expect(result).toMatchSnapshot();
        }
    });

    it('strict:true, allow other errors', () => {
        const dir = path.resolve(__dirname, '__sandbox__/app1/');
        process.chdir(dir);

        const resolve = new ResolveWithPrefix({
            prefix: 'one-preset',
            org: '@example',
            strict: true,
        });

        try {
            expect.assertions(1);
            resolve('one-preset-scoped');
        } catch (error) {
            const result = normalizeRootPath(error);
            expect(result).toMatchSnapshot();
        }
    });

    it('strict:true, do not require module: when prefix not set', () => {
        const dir = path.resolve(__dirname, '__sandbox__/app1/');
        process.chdir(dir);

        const resolve = new ResolveWithPrefix({
            strict: true,
        });

        const resolved = resolve('explicit-name-test');
        const result = normalizeRootPath(resolved);
        expect(result).toMatchSnapshot();
    });

    it('if module not found but it looks like @org missing, add error', () => {
        const dir = path.resolve(__dirname, '__sandbox__/app1/');
        process.chdir(dir);

        const resolve = new ResolveWithPrefix({
            prefix: 'one-preset',
            org: '@example',
        });

        try {
            expect.assertions(1);
            resolve('one-preset-scoped');
        } catch (error) {
            const result = normalizeRootPath(error);
            expect(result).toMatchSnapshot();
        }
    });

    it('works with NODE_PATH', () => {
        const nodePath = process.env.NODE_PATH;
        const dir = path.resolve(__dirname, '__sandbox__/app1/node_modules');

        const sep = platform() === 'win32' ? ';' : ':';
        if (process.env.NODE_PATH) {
            const splitNodePath = process.env.NODE_PATH.split(sep);
            splitNodePath.push(dir);

            process.env.NODE_PATH = splitNodePath.join(':');
        } else {
            process.env.NODE_PATH = dir;
        }

        const resolve = new ResolveWithPrefix({ prefix: 'one-preset' });

        const resolved = resolve('one-preset-test');
        const expected = path.resolve(dir, 'one-preset-test/index.js');

        expect(resolved).toEqual(expected);
        process.env.NODE_PATH = nodePath;
    });
});
