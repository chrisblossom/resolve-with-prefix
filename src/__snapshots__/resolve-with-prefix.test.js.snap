// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResolveWithPrefix default strict:true, require module: 1`] = `
Object {
  "Error": "Cannot find module 'one-preset-explicit-name-test' from '/normalized/root/path'
- If you want to resolve \\"explicit-name-test\\", use \\"module:explicit-name-test\\"",
}
`;

exports[`ResolveWithPrefix if module not found but it looks like @org missing, add error 1`] = `
Object {
  "Error": "Cannot find module 'one-preset-scoped' from '/normalized/root/path'
- Did you mean \\"@example/one-preset-scoped\\"?",
}
`;

exports[`ResolveWithPrefix resolves full preset with prefix set 1`] = `"/normalized/root/path/node_modules/one-preset-test/index.js"`;

exports[`ResolveWithPrefix resolves full preset without prefix set 1`] = `"/normalized/root/path/node_modules/one-preset-test/index.js"`;

exports[`ResolveWithPrefix resolves leading preset with prefix 1`] = `"/normalized/root/path/node_modules/one-preset-test/index.js"`;

exports[`ResolveWithPrefix resolves leading preset with prefix over matching module 1`] = `"/normalized/root/path/node_modules/one-preset-other/index.js"`;

exports[`ResolveWithPrefix resolves other without preset 1`] = `"/normalized/root/path/node_modules/other/index.js"`;

exports[`ResolveWithPrefix strict:true, allow other errors 1`] = `
Object {
  "Error": "Cannot find module 'one-preset-scoped' from '/normalized/root/path'
- Did you mean \\"@example/one-preset-scoped\\"?",
}
`;

exports[`ResolveWithPrefix strict:true, do not require module: when prefix not set 1`] = `"/normalized/root/path/node_modules/explicit-name-test/index.js"`;

exports[`ResolveWithPrefix strict:true, module: works 1`] = `"/normalized/root/path/node_modules/explicit-name-test/index.js"`;

exports[`ResolveWithPrefix strict:true, require module: 1`] = `
Object {
  "Error": "Cannot find module 'one-preset-explicit-name-test' from '/normalized/root/path'
- If you want to resolve \\"explicit-name-test\\", use \\"module:explicit-name-test\\"",
}
`;
